teps/0140-producing-results-in-matrix.md

```diff
@@ -0,0 +1,512 @@
+      taskSpec:
+        params:
+          - name: IMAGE
+          - name: DIGEST
+            default: ""
+        results:
+          - name: IMAGE-DIGEST
+        steps:
+          - name: produce-image-digest
+            image: bash:latest
+            script: |
+              #!/usr/bin/env bash
+              echo "Building image for $(params.IMAGE)"
+              echo -n "$(params.DIGEST)" | sha256sum | tee $(results.IMAGE-DIGEST.path)
+```
+
+
+#### 2. Platforms and Browsers
+
+As a `Pipeline` author, I need to run tests on a combination of platforms and browsers. This will fan out into 9 different `TaskRuns` that will produce a `Result` “report-url” for each combination which can be used in a subsequent `PipelineTask` to fetch the report for each platform-browser combination.
+
+```text
+# platforms
+linux
+windows
+mac
+
+# browsers
+chrome
+firefox
+safari
+```
+
+
+```
+                                                                 clone
+                                                                   |
+                                                                   v
+   --------------------------------------------------------------------------------------------------------------------------
+     |              |              |             |               |                |              |          |          |
+     v              v              v             v               v                v              v          v          v
+linux-chrome  linux-firefox   linux-safari  windows-chrome  windows-firefox  windows-safari  mac-chrome  mac-firefox  mac-safari
+     |              |              |             |               |                |              |          |          |
+     v              v              v             v               v                v              v          v          v
+report-url-0   report-url-1   report-url-2   report-url-3    report-url-4    report-url-5  report-url-6  report-url-7 report-url-8
+```
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: matrix-testing-platform-and-browsers
+spec:
+  tasks:
+    - name: test-platforms-and-browsers
+      matrix:
+        params:
+          - name: platform
+            value:
+              - linux
+              - mac
+              - windows
+          - name: browser
+            value:
+              - chrome
+              - safari
+              - firefox
+      taskSpec:
+        params:
+          - name: platform
+            type: string
+          - name: browser
+            type: string
+        results:
+          - name: report-url
+            type: string
+         steps:
+          - name: produce-report-url
+            image: alpine
+            script: |
+               echo "Running tests on $(params.platform)-$(params.browser)"
+               echo -n "https://api.example/get-report/$(params.platform)-$(params.browser)" | tee $(results.report-url.path)
+```
+
+## Proposal
```

Comment:
What happens if one of the matrixed `Task` fails (prior to write the result) ? Are we producing incomplete results array or no results at all ?

Reply:
In the event that a Matrixed `PipelineTask` that produces `Results` fans out into multiple `TaskRuns` and one or more of the fanned out `TaskRuns` fail, the current behavior of allowing a failed `TaskRun` to still produce a `Result` would be supported https://github.com/tektoncd/pipeline/pull/6510

Reply:
Thanks Vincent for this question, I think it's interesting/important! There are actually a few related questions here IMO:
1. If a matrixed taskrun fails before writing a result, will the array result not exist or be length n - 1 or n (i.e. we provide some "zero value" for the result of a taskrun that doesn't produce a result that should) where n is the number of combinations? I think this is what Vincent was asking. IMO it should be length n - 1, and then once tep 50 is implemented, pipeline authors can choose whether to continue operating on the incomplete result even though the pipeline task is failed.
2. If a matrixed taskrun fails after writing a result, will the array result exist? I think https://github.com/tektoncd/pipeline/pull/6510 allows a taskrun to produce a result even if it failed after writing it, so it seems reasonable to me to have an array result populated for a matrixed pipeline task for consistency.
3. If a matrixed taskrun is supposed to produce a result, but doesn't, and still succeeds, should the array result not exist or be length n - 1 or n? Reading through TEP 48 I'm not sure exactly what the right call is here but I'd say an incomplete array result makes sense for consistency; the pipeline task will succeed so it wouldn't make sense to have it not produce a result.

In general I don't think it would be a good idea to produce "zero value" results for tasks that don't write the results to ensure the array result is the length of the number of combinations. (Note that this is another reason indexing might be a confusing UX: we'd have to either produce results for tasks that don't write them, or not produce an array result any time a task didn't write a result, even if the task succeeded.)